include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h

```diff
@@ -0,0 +1,681 @@
+    const auto& __x = __input[__index];
+    if (__func.is_valid(__x))
+    {
+        _BinIdxType __c = __func.get_bin(__x);
+        __dpl_sycl::__atomic_ref<_histo_value_type, _AddressSpace> __local_bin(__wg_local_histogram[__offset + __c]);
+        ++__local_bin;
+    }
+}
+
+template <typename _BinType, typename _FactorType, typename _HistAccessorIn, typename _OffsetT,
+          typename _HistAccessorOut, typename _Size>
+void
+__reduce_out_histograms(const _HistAccessorIn& __in_histogram, const _OffsetT& __offset,
+                        const _HistAccessorOut& __out_histogram, _Size __num_bins, const sycl::nd_item<1>& __self_item)
+{
+    ::std::uint32_t __gSize = __self_item.get_local_range()[0];
+    ::std::uint32_t __self_lidx = __self_item.get_local_id(0);
+    _FactorType __factor = oneapi::dpl::__internal::__dpl_ceiling_div(__num_bins, __gSize);
+    _FactorType __k = 0;
+
+    for (; __k < __factor - 1; ++__k)
+    {
+        __dpl_sycl::__atomic_ref<_BinType, sycl::access::address_space::global_space> __global_bin(
+            __out_histogram[__gSize * __k + __self_lidx]);
+        __global_bin += __in_histogram[__offset + __gSize * __k + __self_lidx];
+    }
+    // residual
+    if (__gSize * __k + __self_lidx < __num_bins)
+    {
+        __dpl_sycl::__atomic_ref<_BinType, sycl::access::address_space::global_space> __global_bin(
+            __out_histogram[__gSize * __k + __self_lidx]);
+        __global_bin += __in_histogram[__offset + __gSize * __k + __self_lidx];
+    }
+}
+
+template <::std::uint16_t __iters_per_work_item, ::std::uint8_t __bins_per_work_item, typename _KernelName>
+struct __histogram_general_registers_local_reduction_submitter;
+
+template <::std::uint16_t __iters_per_work_item, ::std::uint8_t __bins_per_work_item, typename... _KernelName>
+struct __histogram_general_registers_local_reduction_submitter<__iters_per_work_item, __bins_per_work_item,
+                                                               __internal::__optional_kernel_name<_KernelName...>>
+{
+    template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _BinHashMgr>
+    auto
+    operator()(_ExecutionPolicy&& __exec, const sycl::event& __init_e, ::std::uint16_t __work_group_size,
+               _Range1&& __input, _Range2&& __bins, _BinHashMgr __binhash_manager)
+    {
+        const ::std::size_t __n = __input.size();
+        const ::std::uint8_t __num_bins = __bins.size();
+        using _local_histogram_type = ::std::uint32_t;
+        using _private_histogram_type = ::std::uint16_t;
+        using _histogram_index_type = ::std::uint8_t;
+        using _bin_type = oneapi::dpl::__internal::__value_t<_Range2>;
+        using _extra_memory_type = typename _BinHashMgr::_extra_memory_type;
+        auto _device_copyable_func = __binhash_manager.get_device_copyable_binhash();
+
+        ::std::size_t __extra_SLM_elements = __binhash_manager.get_required_SLM_elements();
```

Comment:
`__binhash_manager_base` is used together with binhash functions which have no extra sequence of memory which requires a buffer to represent them (`evenly_divided_binhash`).  You link to the `custom_range_binhash` specialization of `__binhash_SLM_wrapper`, which has an extra sequence on memory (the boundaries) which need to be held in a buffer, and will therefore us `__binhash_manager_with_buffer` instead.
No out of range access will occur here.

It would be nice if `binhash_manager_...` and `__binhash_SLM_wrapper` could be combined, but the first needs to be set up in host code and keep alive a `sycl::buffer` for the duration of the kernel, and the other needs to be created in the kernel to be able to be created with an SLM accessor, and accelerate the kernel without passing around SLM accessors to each call.

Reply:
remove me